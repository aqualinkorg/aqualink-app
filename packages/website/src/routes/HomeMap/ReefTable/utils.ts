import { startCase } from "lodash";
import type { TableRow } from "../../../store/Homepage/types";

export type Order = "asc" | "desc";

export enum OrderKeys {
  ALERT = "alert",
  DEPTH = "depth",
  DHW = "dhw",
  LOCATION_NAME = "locationName",
  TEMP = "temp",
}

// This type isn't used anywhere, it just forces the above enum to only hold valid keys to TableRow.
type __CheckOrderKeys = TableRow[OrderKeys];

/**
 * Returns a human friendly string of a order key.
 * @param orderKey key to return the string for.
 */
export function getOrderKeysFriendlyString(orderKey: OrderKeys) {
  switch (orderKey) {
    case OrderKeys.DHW:
      return "DHW";
    default:
      return startCase(orderKey);
  }
}

export function descendingComparator<T>(a: T, b: T, orderBy: keyof T) {
  return b[orderBy] >= a[orderBy] ? 1 : -1;
}

export function getComparator(
  order: Order,
  orderBy: OrderKeys
): (
  a: {
    [key in OrderKeys]: number | string | null;
  },
  b: {
    [key in OrderKeys]: number | string | null;
  }
) => number {
  return order === "desc"
    ? (a, b) => descendingComparator(a, b, orderBy)
    : (a, b) => -descendingComparator(a, b, orderBy);
}

export function stableSort<T>(array: T[], comparator: (a: T, b: T) => number) {
  // eslint-disable-next-line fp/no-mutating-methods
  return [...array].sort(comparator);
}
