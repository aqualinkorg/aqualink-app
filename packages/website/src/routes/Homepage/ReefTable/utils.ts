export type Order = "asc" | "desc";

export type OrderKeys = "locationName" | "temp" | "depth" | "dhw" | "alert";

export function descendingComparator<T>(a: T, b: T, orderBy: keyof T) {
  return b[orderBy] >= a[orderBy] ? 1 : -1;
}

export function getComparator(
  order: Order,
  orderBy: OrderKeys
): (
  a: {
    [key in OrderKeys]: number | string | null;
  },
  b: {
    [key in OrderKeys]: number | string | null;
  }
) => number {
  return order === "desc"
    ? (a, b) => descendingComparator(a, b, orderBy)
    : (a, b) => -descendingComparator(a, b, orderBy);
}

export function stableSort<T>(array: T[], comparator: (a: T, b: T) => number) {
  return [...array].sort(comparator);
}
