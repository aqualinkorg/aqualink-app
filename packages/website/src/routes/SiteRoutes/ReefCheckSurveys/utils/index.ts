type SegmentsEntity = {
  s1: number;
  s2: number;
  s3: number;
  s4: number;
};

export function segmentsTotalSortComparator<T extends SegmentsEntity>(
  a: T,
  b: T,
) {
  return b.s1 + b.s2 + b.s3 + b.s4 - (a.s1 + a.s2 + a.s3 + a.s4);
}
