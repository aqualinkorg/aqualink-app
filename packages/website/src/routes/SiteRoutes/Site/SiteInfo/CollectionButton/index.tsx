import React, { useEffect, useState } from 'react';
import { Tooltip, IconButton, useTheme, Theme } from '@mui/material';
import { WithStyles } from '@mui/styles';
import createStyles from '@mui/styles/createStyles';
import withStyles from '@mui/styles/withStyles';
import { useDispatch, useSelector } from 'react-redux';

import {
  createCollectionRequest,
  setCollectionSites,
  userCollectionLoadingSelector,
  userErrorSelector,
  userInfoSelector,
} from 'store/User/userSlice';
import { belongsToCollection } from 'helpers/siteUtils';
import collectionServices from 'services/collectionServices';
import { ReactComponent as WatchIcon } from 'assets/watch.svg';
import { ReactComponent as UnWatchIcon } from 'assets/unwatch.svg';

const CollectionButton = ({
  siteId,
  errorCallback,
  classes,
}: CollectionButtonProps) => {
  const theme = useTheme();
  const dispatch = useDispatch();
  const user = useSelector(userInfoSelector);
  const userCollectionLoading = useSelector(userCollectionLoadingSelector);
  const userError = useSelector(userErrorSelector);
  const [collectionActionLoading, setCollectionActionLoading] = useState(false);
  const siteBelongsToCollection = belongsToCollection(
    siteId,
    user?.collection?.siteIds,
  );

  useEffect(() => {
    if (userError) errorCallback();
  }, [errorCallback, userError]);

  useEffect(() => {
    if (!userCollectionLoading) {
      setCollectionActionLoading(false);
    }
  }, [userCollectionLoading]);

  const onAddSiteToCollection = () => {
    if (user?.token && !user.collection) {
      setCollectionActionLoading(true);
      dispatch(
        createCollectionRequest({
          name: `${user.fullName}'s collection`,
          siteIds: [siteId],
          token: user.token,
        }),
      );
    } else if (user?.token && user?.collection && !siteBelongsToCollection) {
      setCollectionActionLoading(true);
      collectionServices
        .updateCollection(
          {
            id: user.collection.id,
            addSiteIds: [siteId],
          },
          user.token,
        )
        .then(() => {
          if (user?.collection) {
            dispatch(setCollectionSites([...user.collection.siteIds, siteId]));
          }
        })
        .catch(() => errorCallback())
        .finally(() => setCollectionActionLoading(false));
    }
  };

  const onRemoveSiteFromCollection = () => {
    if (user?.token && user?.collection && siteBelongsToCollection) {
      setCollectionActionLoading(true);
      collectionServices
        .updateCollection(
          {
            id: user.collection.id,
            removeSiteIds: [siteId],
          },
          user.token,
        )
        .then(() => {
          if (user?.collection) {
            dispatch(
              setCollectionSites(
                user.collection.siteIds.filter((item) => item !== siteId),
              ),
            );
          }
        })
        .catch(() => errorCallback())
        .finally(() => setCollectionActionLoading(false));
    }
  };

  return (
    <Tooltip
      title={
        siteBelongsToCollection
          ? 'Remove from your dashboard'
          : 'Add to your dashboard'
      }
      arrow
      placement="top"
    >
      <IconButton
        className={classes.root}
        disabled={collectionActionLoading}
        onClick={
          siteBelongsToCollection
            ? onRemoveSiteFromCollection
            : onAddSiteToCollection
        }
        size="large"
      >
        {siteBelongsToCollection ? (
          <UnWatchIcon
            color={
              collectionActionLoading
                ? theme.palette.grey[500]
                : theme.palette.error.main
            }
          />
        ) : (
          <WatchIcon
            color={
              collectionActionLoading
                ? theme.palette.grey[500]
                : theme.palette.primary.main
            }
          />
        )}
      </IconButton>
    </Tooltip>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    root: {
      margin: theme.spacing(-2.5, -1.5, -1, 0.5),
    },
  });

interface CollectionIncomingButtonProps {
  siteId: number;
  errorCallback: () => void;
}

type CollectionButtonProps = CollectionIncomingButtonProps &
  WithStyles<typeof styles>;

export default withStyles(styles)(CollectionButton);
