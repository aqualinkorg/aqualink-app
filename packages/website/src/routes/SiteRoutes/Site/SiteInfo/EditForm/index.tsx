import React, { ChangeEvent, FormEvent } from 'react';
import {
  Button,
  Grid,
  Typography,
  Checkbox,
  FormControlLabel,
} from '@mui/material';
import { WithStyles } from '@mui/styles';
import withStyles from '@mui/styles/withStyles';
import createStyles from '@mui/styles/createStyles';
import Alert from '@mui/material/Alert';
import { useDispatch, useSelector } from 'react-redux';
import { find } from 'lodash';

import { Site, SiteUpdateParams, Status } from 'store/Sites/types';
import { getSiteNameAndRegion } from 'store/Sites/helpers';
import { siteDraftSelector, setSiteDraft } from 'store/Sites/selectedSiteSlice';
import { useFormField } from 'hooks/useFormField';
import TextField from 'common/Forms/TextField';
import StatusSelector from 'common/StatusSelector';
import { userInfoSelector } from 'store/User/userSlice';
import { sanitizeUrl } from '@braintree/sanitize-url';

const NUMERIC_FIELD_STEP = 1 / 10 ** 15;

const EditForm = ({
  site,
  loading,
  onClose,
  onSubmit,
  classes,
}: EditFormProps) => {
  const dispatch = useDispatch();
  const user = useSelector(userInfoSelector);

  const draftSite = useSelector(siteDraftSelector);
  const location = site.polygon.type === 'Point' ? site.polygon : null;
  const { latitude: draftLatitude, longitude: draftLongitude } =
    draftSite?.coordinates || {};

  const [editToken, setEditToken] = React.useState(
    !!site.maskedSpotterApiToken,
  );

  const [editContactInfo, setEditContactInfo] = React.useState(false);

  const setDraftSiteCoordinates =
    (field: 'longitude' | 'latitude') => (value: string) => {
      dispatch(
        setSiteDraft({
          ...draftSite,
          coordinates: draftSite?.coordinates && {
            ...draftSite.coordinates,
            [field]: parseFloat(value),
          },
        }),
      );
    };

  // Form Fields
  const [siteName, setSiteName] = useFormField<string>(
    getSiteNameAndRegion(site).name ?? '',
    ['required', 'maxLength'],
  );

  const [siteDepth, setSiteDepth] = useFormField<string>(
    site.depth?.toString() ?? '',
    ['required', 'isInt'],
  );

  const [siteLatitude, setSiteLatitude] = useFormField<string>(
    location?.coordinates[1].toString() ?? '',
    ['required', 'isNumeric', 'isLat'],
    draftLatitude?.toString(),
    setDraftSiteCoordinates('latitude'),
  );

  const [siteLongitude, setSiteLongitude] = useFormField<string>(
    location?.coordinates[0].toString() ?? '',
    ['required', 'isNumeric', 'isLong'],
    draftLongitude?.toString(),
    setDraftSiteCoordinates('longitude'),
  );

  const [siteSensorId, setSensorId] = useFormField<string>(
    site.sensorId ?? '',
    ['maxLength'],
  );

  const [siteSpotterApiToken, setSiteSpotterApiToken] = useFormField<string>(
    site.maskedSpotterApiToken ?? '',
    ['maxLength'],
  );

  const [apiTokenChanged, setApiTokenChanged] = React.useState(false);

  const [status, setStatus] = useFormField<string>(site.status, []);

  const [display, setDisplay] = useFormField<boolean>(site.display, []);

  const [videoStream, setVideoStream] = useFormField<string>(
    site.videoStream || '',
    [],
  );

  const [contactInformation, setContactInformation] = useFormField<string>(
    site.contactInformation || '',
    ['maxLength'],
  );

  const [siteIframe, setSiteIframe] = useFormField<string>(site.iframe || '', [
    'maxLength',
  ]);

  const onApiTokenBlur = () => {
    if (siteSpotterApiToken.value === '' && site.maskedSpotterApiToken) {
      setSiteSpotterApiToken(site.maskedSpotterApiToken);
      setApiTokenChanged(false);
    }
  };
  const onApiTokenFocus = () => {
    if (!apiTokenChanged) {
      setSiteSpotterApiToken('');
    }
  };

  const formSubmit = (event: FormEvent<HTMLFormElement>) => {
    if (
      siteName.value &&
      siteDepth.value &&
      siteLatitude.value &&
      siteLongitude.value
    ) {
      const insertedTokenValue = apiTokenChanged
        ? siteSpotterApiToken.value
        : undefined;
      const spotterApiToken = editToken ? insertedTokenValue : null;
      // fields need to be undefined in order not be affected by the update.
      // siteSensorId.value here can be <empty string> which our api does not accept
      const sensorId = siteSensorId.value || undefined;
      const statusVal = status.value;
      const displayVal = display.value;
      const videoStreamVal = videoStream.value || null;
      const contactInformationVal = contactInformation.value || null;
      const updateParams: SiteUpdateParams = {
        coordinates: {
          latitude: parseFloat(siteLatitude.value),
          longitude: parseFloat(siteLongitude.value),
        },
        name: siteName.value,
        depth: parseInt(siteDepth.value, 10),
        sensorId,
        spotterApiToken,
        ...(siteIframe.value ? { iframe: sanitizeUrl(siteIframe.value) } : {}),
        ...(user?.adminLevel === 'super_admin' && statusVal
          ? { status: statusVal as Status }
          : {}),
        ...(user?.adminLevel === 'super_admin' ? { display: displayVal } : {}),
        ...(user?.adminLevel === 'super_admin'
          ? { videoStream: videoStreamVal }
          : {}),
        ...(user?.adminLevel === 'super_admin' && editContactInfo
          ? { contactInformation: contactInformationVal }
          : {}),
      };
      onSubmit(updateParams);
    }
    event.preventDefault();
  };

  const onFieldChange = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    checked?: boolean,
  ) => {
    const { name: field, value: newValue } = event.target;
    switch (field) {
      case 'siteName':
        setSiteName(newValue);
        break;
      case 'depth':
        setSiteDepth(newValue);
        break;
      case 'latitude':
        setSiteLatitude(newValue, true);
        break;
      case 'longitude':
        setSiteLongitude(newValue, true);
        break;
      case 'spotterId':
        setSensorId(newValue);
        break;
      case 'spotterApiToken':
        setSiteSpotterApiToken(newValue.trim());
        setApiTokenChanged(true);
        break;
      case 'status':
        setStatus(newValue);
        break;
      case 'display':
        setDisplay(!!checked);
        break;
      case 'videoStream':
        setVideoStream(newValue);
        break;
      case 'contactInformation':
        setContactInformation(newValue);
        break;
      case 'iframe':
        setSiteIframe(newValue);
        break;
      default:
        break;
    }
  };

  return (
    <form onSubmit={formSubmit}>
      <Grid container alignItems="flex-end" spacing={3}>
        <Grid container item sm={12} md={6} spacing={2}>
          <Grid item sm={8} xs={12}>
            <TextField
              formField={siteName}
              label="Site Name"
              placeholder="Site Name"
              name="siteName"
              onChange={onFieldChange}
            />
          </Grid>
          <Grid item sm={4} xs={12}>
            <TextField
              formField={siteDepth}
              label="Depth (m)"
              placeholder="Depth (m)"
              name="depth"
              isNumeric
              step={1}
              onChange={onFieldChange}
            />
          </Grid>
          <Grid item xs={12}>
            <Alert className={classes.infoAlert} icon={false} severity="info">
              <Typography variant="subtitle2">
                Only Sofar Ocean spotters are supported for now. Please contact
                us if you would like to connect other live buoys at{' '}
                <a
                  href="mailto:admin@aqualink.org?subject=Connect%20a%20buoy"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  admin@aqualink.org
                </a>
                .
              </Typography>
            </Alert>
          </Grid>
          <Grid item sm={8} xs={8}>
            <TextField
              formField={siteSensorId}
              label="Spotter ID"
              placeholder="Spotter ID"
              name="spotterId"
              onChange={onFieldChange}
            />
          </Grid>
          <Grid
            item
            xs={4}
            style={{ display: 'flex', justifyContent: 'flex-end' }}
          >
            <FormControlLabel
              labelPlacement="start"
              control={
                <Checkbox
                  color="primary"
                  checked={editToken}
                  onChange={() => setEditToken(!editToken)}
                />
              }
              label="Add a custom API token"
            />
          </Grid>
          {editToken && (
            <Grid item xs={12}>
              <TextField
                formField={siteSpotterApiToken}
                label="Spotter API token"
                placeholder="Spotter API token"
                name="spotterApiToken"
                onChange={onFieldChange}
                onBlur={onApiTokenBlur}
                onFocus={onApiTokenFocus}
              />
            </Grid>
          )}
          <Grid item xs={12}>
            <Alert className={classes.infoAlert} icon={false} severity="info">
              <Typography variant="subtitle2">
                You can also change your site position by dragging the pin on
                the map.
              </Typography>
            </Alert>
          </Grid>
          <Grid item sm={6} xs={12}>
            <TextField
              formField={siteLatitude}
              label="Latitude"
              placeholder="Latitude"
              name="latitude"
              isNumeric
              step={NUMERIC_FIELD_STEP}
              onChange={onFieldChange}
            />
          </Grid>
          <Grid item sm={6} xs={12}>
            <TextField
              formField={siteLongitude}
              label="Longitude"
              placeholder="Longitude"
              name="longitude"
              isNumeric
              step={NUMERIC_FIELD_STEP}
              onChange={onFieldChange}
            />
          </Grid>
          {user?.adminLevel === 'super_admin' && (
            <Grid item xs={4}>
              <StatusSelector
                status={status.value as Status | ''}
                onChange={onFieldChange}
                name="status"
              />
            </Grid>
          )}
          {user?.adminLevel === 'super_admin' && (
            <Grid
              item
              xs={2}
              style={{ display: 'flex', justifyContent: 'flex-end' }}
            >
              <FormControlLabel
                labelPlacement="start"
                control={
                  <Checkbox
                    color="primary"
                    name="display"
                    checked={display.value}
                    onChange={onFieldChange}
                  />
                }
                label="display"
              />
            </Grid>
          )}
          {user?.adminLevel === 'super_admin' && (
            <Grid item xs={6}>
              <TextField
                formField={videoStream}
                label="Video Stream"
                placeholder="Video Stream"
                name="videoStream"
                onChange={onFieldChange}
              />
            </Grid>
          )}
          {user?.adminLevel === 'super_admin' && (
            <Grid
              item
              xs={4}
              style={{ display: 'flex', justifyContent: 'flex-start' }}
            >
              <FormControlLabel
                labelPlacement="start"
                control={
                  <Checkbox
                    color="primary"
                    checked={editContactInfo}
                    onChange={() => setEditContactInfo(!editContactInfo)}
                  />
                }
                label="Edit Contact Info"
              />
            </Grid>
          )}
          {user?.adminLevel === 'super_admin' && (
            <Grid item xs={8}>
              <TextField
                disabled={!editContactInfo}
                formField={contactInformation}
                label="Contact Information"
                placeholder="Contact Information"
                name="contactInformation"
                onChange={onFieldChange}
              />
            </Grid>
          )}
          {
            // Temporally hide this option
            false && (
              <TextField
                formField={siteIframe}
                label="iframe"
                placeholder="iframe"
                name="iframe"
                onChange={onFieldChange}
              />
            )
          }
        </Grid>
        <Grid
          container
          justifyContent="flex-end"
          item
          sm={12}
          md={4}
          spacing={3}
        >
          <Grid item>
            <Button
              onClick={onClose}
              variant="outlined"
              size="medium"
              color="secondary"
            >
              Cancel
            </Button>
          </Grid>
          <Grid item>
            <Button
              type="submit"
              disabled={
                loading ||
                !!find(
                  [siteName, siteDepth, siteLongitude, siteLatitude],
                  (field) => field.error,
                )
              }
              variant="outlined"
              size="medium"
              color="primary"
            >
              {loading ? 'Saving...' : 'Save'}
            </Button>
          </Grid>
        </Grid>
      </Grid>
    </form>
  );
};

const styles = () =>
  createStyles({
    textField: {
      color: 'black',
      alignItems: 'center',
    },
    infoAlert: {
      marginTop: '0.5rem',
    },
  });

interface EditFormIncomingProps {
  site: Site;
  loading: boolean;
  onClose: () => void;
  onSubmit: (data: SiteUpdateParams) => void;
}

type EditFormProps = EditFormIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(EditForm);
