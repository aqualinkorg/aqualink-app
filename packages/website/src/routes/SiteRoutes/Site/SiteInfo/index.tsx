import React, { useState } from "react";
import {
  Grid,
  Typography,
  IconButton,
  withStyles,
  WithStyles,
  createStyles,
  Button,
  Collapse,
  useMediaQuery,
  useTheme,
} from "@material-ui/core";
import Alert from "@material-ui/lab/Alert";
import CloseIcon from "@material-ui/icons/Close";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import ArrowBack from "@material-ui/icons/ArrowBack";

import EditForm from "./EditForm";
import ExclusionDatesDialog from "./ExclusionDatesDialog";
import {
  setSelectedSite,
  setSiteData,
  setSiteDraft,
} from "../../../../store/Sites/selectedSiteSlice";
import { Site, SiteUpdateParams } from "../../../../store/Sites/types";
import { getSiteNameAndRegion } from "../../../../store/Sites/helpers";
import siteServices from "../../../../services/siteServices";
import {
  setAdministeredSiteName,
  userInfoSelector,
} from "../../../../store/User/userSlice";
import { displayTimeInLocalTimezone } from "../../../../helpers/dates";
import CollectionButton from "./CollectionButton";
import {
  sitesListSelector,
  setSiteName,
} from "../../../../store/Sites/sitesListSlice";

const SiteNavBar = ({
  hasDailyData,
  site,
  lastSurvey,
  isAdmin,
  classes,
}: SiteNavBarProps) => {
  const dispatch = useDispatch();
  const theme = useTheme();
  const user = useSelector(userInfoSelector);
  const sitesList = useSelector(sitesListSelector);
  const [editEnabled, setEditEnabled] = useState<boolean>(false);
  const [alertOpen, setAlertOpen] = useState<boolean>(false);
  const [alertSeverity, setAlertSeverity] = useState<"success" | "error">();
  const [formSubmitLoading, setFormSubmitLoading] = useState(false);
  const { name: siteName, region: siteRegion } = getSiteNameAndRegion(site);
  const organizationName = site.admins[0]?.organization;
  const matches = useMediaQuery(theme.breakpoints.down("sm"));
  const [exclusionDatesDialogOpen, setExclusionDatesDeployDialogOpen] =
    useState(false);

  const clearSiteInfo = () => {
    if (!hasDailyData) {
      dispatch(setSelectedSite(null));
    }
  };

  const onCloseForm = () => {
    dispatch(setSiteDraft(null));
    setEditEnabled(false);
  };

  const onOpenForm = () => {
    if (site.depth && site.polygon.type === "Point") {
      dispatch(
        setSiteDraft({
          name: siteName || "",
          depth: site.depth,
          coordinates: {
            longitude: site.polygon.coordinates[0],
            latitude: site.polygon.coordinates[1],
          },
        })
      );
    }
    setEditEnabled(true);
  };

  const handleFormSubmit = (data: SiteUpdateParams) => {
    if (user && user.token) {
      setFormSubmitLoading(true);
      siteServices
        .updateSite(site.id, data, user.token)
        .then(() => dispatch(setSiteData(data)))
        .then(() => {
          dispatch(
            setAdministeredSiteName({
              id: site.id,
              list: user?.administeredSites,
              name: data.name,
            })
          );
          dispatch(
            setSiteName({
              id: site.id,
              list: sitesList,
              name: data.name,
            })
          );
        })
        .then(() => setAlertSeverity("success"))
        .catch(() => setAlertSeverity("error"))
        .finally(() => {
          dispatch(setSiteDraft(null));
          setEditEnabled(false);
          setAlertOpen(true);
          setFormSubmitLoading(false);
        });
    }
  };

  return (
    <>
      {user?.token &&
        isAdmin &&
        (site.status === "shipped" || site.status === "deployed") && (
          <ExclusionDatesDialog
            dialogType={site.status === "shipped" ? "deploy" : "maintain"}
            open={exclusionDatesDialogOpen}
            onClose={() => setExclusionDatesDeployDialogOpen(false)}
            token={user.token}
            timeZone={site.timezone}
            siteId={site.id}
          />
        )}
      <Collapse in={alertOpen}>
        <Alert
          severity={alertSeverity}
          action={
            <IconButton
              color="inherit"
              size="small"
              onClick={() => {
                setAlertOpen(false);
              }}
            >
              <CloseIcon fontSize="inherit" />
            </IconButton>
          }
        >
          {alertSeverity === "success"
            ? "Successfully updated site information"
            : "Something went wrong"}
        </Alert>
      </Collapse>
      <Grid
        className={classes.root}
        container
        justify="space-between"
        alignItems="center"
      >
        <Grid item xs={12}>
          <Grid alignItems="center" container spacing={1}>
            {!editEnabled ? (
              <Grid item className={classes.headerButtonWrapper}>
                <IconButton
                  onClick={clearSiteInfo}
                  edge="start"
                  color="primary"
                  aria-label="menu"
                  component={Link}
                  to="/map"
                >
                  <ArrowBack />
                </IconButton>
              </Grid>
            ) : null}

            {editEnabled ? (
              <Grid item xs={12}>
                <EditForm
                  site={site}
                  loading={formSubmitLoading}
                  onClose={onCloseForm}
                  onSubmit={handleFormSubmit}
                />
              </Grid>
            ) : (
              <Grid
                className={classes.headerWrapper}
                container
                alignItems="center"
                item
                spacing={1}
              >
                <Grid item xs={12} md={8} container alignItems="center">
                  <Grid item xs={12}>
                    <Grid container alignItems="baseline">
                      <Grid className={classes.siteNameWrapper} item>
                        <Typography variant="h4">
                          {siteName}
                          {siteRegion && `, ${siteRegion}`}
                        </Typography>
                      </Grid>
                      <Grid className={classes.headerButtonWrapper} item>
                        <CollectionButton
                          siteId={site.id}
                          errorCallback={() => {
                            setAlertOpen(true);
                            setAlertSeverity("error");
                          }}
                        />
                      </Grid>
                    </Grid>
                  </Grid>
                  {organizationName && (
                    <Grid item xs={12}>
                      <Typography variant="h6">{`Managed by ${organizationName}`}</Typography>
                    </Grid>
                  )}
                  {lastSurvey && (
                    <Grid item xs={12}>
                      <Typography variant="subtitle1">{`Last surveyed: ${displayTimeInLocalTimezone(
                        {
                          isoDate: lastSurvey,
                          format: "MMM DD[,] YYYY",
                          displayTimezone: false,
                          timeZone: site.timezone,
                        }
                      )}`}</Typography>
                    </Grid>
                  )}
                </Grid>
                {isAdmin && (
                  <Grid
                    container
                    direction={matches ? "row" : "column"}
                    item
                    xs={12}
                    md={4}
                    spacing={1}
                  >
                    <Grid item>
                      <Button
                        className={classes.button}
                        onClick={onOpenForm}
                        size="small"
                        color="primary"
                        variant="outlined"
                      >
                        EDIT SITE DETAILS
                      </Button>
                    </Grid>
                    {site.sensorId &&
                      (site.status === "shipped" ||
                        site.status === "deployed") && (
                        <Grid item>
                          <Button
                            className={classes.button}
                            onClick={() =>
                              setExclusionDatesDeployDialogOpen(true)
                            }
                            size="small"
                            color="primary"
                            variant="outlined"
                          >
                            {site.status === "shipped"
                              ? "MARK AS DEPLOYED"
                              : "ADD EXCLUSION DATES"}
                          </Button>
                        </Grid>
                      )}
                    <Grid item>
                      <Button
                        component={Link}
                        to={`/sites/${site.id}/upload_data`}
                        className={classes.button}
                        color="primary"
                        variant="outlined"
                        size="small"
                      >
                        UPLOAD DATA
                      </Button>
                    </Grid>
                  </Grid>
                )}
              </Grid>
            )}
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};

const styles = () =>
  createStyles({
    root: {
      marginTop: "2rem",
    },
    managerInfo: {
      marginRight: "0.5rem",
    },
    button: {
      minWidth: 180,
    },
    headerButtonWrapper: {
      width: 48,
    },
    headerWrapper: {
      maxWidth: "calc(100% - 48px)", // maximum width of 100% minus the width of the back button
    },
    siteNameWrapper: {
      maxWidth: "calc(100% - 56px)", // maximum width of 100% minus the width of the collection button and its padding
      overflowWrap: "break-word",
    },
  });

interface SiteNavBarIncomingProps {
  hasDailyData: boolean;
  site: Site;
  lastSurvey?: string | null;
  isAdmin: boolean;
}

SiteNavBar.defaultProps = {
  lastSurvey: null,
};

type SiteNavBarProps = SiteNavBarIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(SiteNavBar);
