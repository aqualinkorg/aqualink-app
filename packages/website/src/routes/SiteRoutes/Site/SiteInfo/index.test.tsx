import React from 'react';
import configureStore from 'redux-mock-store';

import { mockSite } from 'mocks/mockSite';
import { mockUser } from 'mocks/mockUser';
import { renderWithProviders } from 'utils/test-utils';
import SiteInfo from '.';

const mockStore = configureStore([]);

describe('SiteInfo', () => {
  let element: HTMLElement;
  beforeEach(() => {
    const store = mockStore({
      user: {
        userInfo: mockUser,
      },
      sitesList: {
        list: [mockSite],
      },
      selectedSite: {
        contactInfoLoading: false,
        loading: false,
        error: null,
      },
    });
    store.dispatch = vi.fn();

    element = renderWithProviders(
      <SiteInfo
        site={mockSite}
        isAdmin
        lastSurvey="2020-09-10T10:27:00.000Z"
      />,
      { store },
    ).container;
  });

  it('should render with given state from Redux store', () => {
    expect(element).toMatchSnapshot();
  });
});
