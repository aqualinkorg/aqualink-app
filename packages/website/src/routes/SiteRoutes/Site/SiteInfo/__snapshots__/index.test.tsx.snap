// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteNavBar should render with given state from Redux store 1`] = `
<div>
  <div
    class="MuiCollapse-container MuiCollapse-hidden"
    style="min-height: 0px;"
  >
    <div
      class="MuiCollapse-wrapper"
    >
      <div
        class="MuiCollapse-wrapperInner"
      >
        <div
          class="MuiPaper-root MuiAlert-root MuiAlert-standardSuccess MuiPaper-elevation0"
          role="alert"
        >
          <div
            class="MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2, 4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0, 0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message"
          >
            Something went wrong
          </div>
          <div
            class="MuiAlert-action"
          >
            <mock-iconbutton
              color="inherit"
              size="small"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </mock-iconbutton>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root SiteNavBar-root-1 MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
      >
        <div
          class="MuiGrid-root SiteNavBar-headerButtonWrapper-4 MuiGrid-item"
        >
          <mock-iconbutton
            aria-label="menu"
            color="primary"
            component="[object Object]"
            edge="start"
            to="/map"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
          </mock-iconbutton>
        </div>
        <div
          class="MuiGrid-root SiteNavBar-headerWrapper-5 MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-item MuiGrid-align-items-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-grid-xs-12 MuiGrid-grid-md-8"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-baseline"
              >
                <div
                  class="MuiGrid-root SiteNavBar-siteNameWrapper-6 MuiGrid-item"
                >
                  <mock-typography
                    variant="h4"
                  >
                    Hawaii
                  </mock-typography>
                </div>
                <div
                  class="MuiGrid-root SiteNavBar-headerButtonWrapper-4 MuiGrid-item"
                />
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <mock-typography
                variant="h6"
              >
                Managed by Some organization
              </mock-typography>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <mock-typography
                variant="subtitle1"
              >
                Last surveyed: Sep 10, 2020
              </mock-typography>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <mock-button
                classname="SiteNavBar-button-3"
                color="primary"
                size="small"
                variant="outlined"
              >
                EDIT SITE DETAILS
              </mock-button>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <mock-button
                classname="SiteNavBar-button-3"
                color="primary"
                component="[object Object]"
                size="small"
                to="/sites/1/upload_data"
                variant="outlined"
              >
                UPLOAD DATA
              </mock-button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
